update doc about codebase deployment (#1051)
open-mmlab/mmdeploy: docs/en/04-supported-codebases/mmdet.md

# MMDetection Deployment

[MMDetection](https://github.com/open-mmlab/mmdetection) aka `mmdet` is an open source object detection toolbox based on PyTorch. It is a part of the [OpenMMLab](https://openmmlab.com/) project.

## Installation

### Install mmdet

If you have already done that, move on to [the next section.](#install-mmdeploy) Otherwise, please follow the [guide](https://mmdetection.readthedocs.io/en/3.x/get_started.html) to finish mmdet installation.

### Install mmdeploy

There are several methods to install mmdeploy, among which you can choose an appropriate one according to your target platform and device.

**Method I:** Install precompiled package

> **TODO**. MMDeploy hasn't released based on dev-1.x branch.

**Method II:** Build using scripts

If your target platform is **Ubuntu 18.04 or later version**, we encourage you to run

> [scripts](../01-how-to-build/build_from_script.md). For example, the following commands help install mmdeploy as well as inference engine - `ONNX Runtime` automatically.

Reviewer: For example, the following commands help install mmdeploy --》For example, the following commands install mmdeploy and the inference engine -`ONNX Runtime`.

```shell
git clone --recursive -b dev-1.x https://github.com/open-mmlab/mmdeploy.git
cd mmdeploy
python3 tools/scripts/build_ubuntu_x64_ort.py $(nproc)
export PYTHONPATH=$(pwd)/build/lib:$PYTHONPATH
export LD_LIBRARY_PATH=$(pwd)/../mmdeploy-dep/onnxruntime-linux-x64-1.8.1/lib/:$LD_LIBRARY_PATH
```

**Method III:** Build from source

If neither **I** nor **II** meets your requirements, [building mmdeploy from source](../01-how-to-build/build_from_source.md) is the last option.

## Convert model

You can use [tools/deploy.py](https://github.com/open-mmlab/mmdeploy/blob/dev-1.x/tools/deploy.py) to convert mmdet models to the specified backend models. Its detailed usage can be learned from [here](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/02-how-to-run/convert_model.md#usage).

> The following shows an example about converting `faster r-cnn` model to onnx model that can be inferred by ONNX Runtime.

Reviewer: The command below or The following command

  mmdet models like `RetinaNet`, `Faster R-CNN` and `DETR` and so on belongs to `detection` task. While `Mask R-CNN` is one of `instance-seg` models. You can find more of them in chapter [Supported models](#supported-models).

  **DO remember to use** `detection/detection_*.py` deployment config file when trying to converting detection models and use `instance-seg/instance-seg_*.py` to deploy instance segmentation models.

- **{backend}:** inference backend, such as onnxruntime, tensorrt, pplnn, ncnn, openvino, coreml and etc.

- **{precision}:** fp16, int8. When it's empty, it means fp32

- **{static | dynamic}:** static shape or dynamic shape

- **{shape}:** input shape or shape range of a model

Therefore, in the above example, you can also convert `faster r-cnn` to other backend models by changing the deployment config file `detection_onnxruntime_dynamic.py` to [others](https://github.com/open-mmlab/mmdeploy/tree/dev-1.x/configs/mmdet/detection), e.g., converting to tensorrt-fp16 model by `detection_tensorrt-fp16_dynamic-320x320-1344x1344.py`.

```{tip}
When converting mmdet models to tensorrt models, --device should be set to "cuda"
```

## Model Specification

> Before moving on to model inference chapter, let us talk more about the converted model structure which is very important to do model inference.

Reviewer: let us talk more about the converted model structure which is very important to do model inference. -> let's know more about the converted model structure which is important for model inference.

> The converted model locates in the working directory like `mmdeploy_models/mmdet/ort` in the previous example. It includes:

Reviewer: It includes: -> The output directory contains four files shown as below

```
mmdeploy_models/mmdet/ort
├── deploy.json
├── detail.json
├── end2end.onnx
└── pipeline.json
```

> in which,

Reviewer: Each JSON file contains the necessary information for model inference:

- **end2end.onnx**: backend model which can be inferred by ONNX Runtime
- **deploy.json**: meta information about backend model
- **pipeline.json**: inference pipeline of mmdeploy SDK
- **detail.json**: conversion parameters

> And the whole package **mmdeploy_models/mmdet/ort** is defined as **mmdeploy SDK model**. In other words, **mmdeploy SDK model** includes not only backend model but also inference meta information.

Reviewer: And the -> The

> And the whole package **mmdeploy_models/mmdet/ort** is defined as **mmdeploy SDK model**. In other words, **mmdeploy SDK model** includes not only backend model but also inference meta information.

Reviewer: ```suggestion
The whole package **mmdeploy_models/mmdet/ort** is defined as **mmdeploy SDK model**, i.e., **mmdeploy SDK model** includes both backend model and inference meta information.
```

## Backend model inference

> mmdeploy provides a unified API named as `inference_model` to do this job, making all inference backends API transparent to users.

Reviewer: MMDeploy

Reviewer: do this job -> inference model

> Take the previous converted `end2end.onnx` model as an example,

Reviewer: Take ..... as an example, you can use the following code to inference the model and visualize the results.

device = 'cpu'
backend_model = ['./mmdeploy_models/mmdet/ort/end2end.onnx']
image = './tests/data/tiger.jpeg'

# read deploy_cfg and model_cfg
deploy_cfg, model_cfg = load_config(deploy_cfg, model_cfg)

# build task and backend model
task_processor = build_task_processor(model_cfg, deploy_cfg, device)
model = task_processor.build_backend_model(backend_model)

# process input image
input_shape = get_input_shape(deploy_cfg)
model_inputs, _ = task_processor.create_input(image, input_shape)

# do model inference
with torch.no_grad():
    result = model.test_step(model_inputs)

# visualize results
task_processor.visualize(
    image=image,
    model=model,
    result=result[0],
    window_name='visualize',
    output_file='output_detection.png')
```

## SDK model inference

TODO

## Supported models

|                                              Model                                               |         Task         | OnnxRuntime | TensorRT | ncnn | PPLNN | OpenVINO |
| :----------------------------------------------------------------------------------------------: | :------------------: | :---------: | :------: | :--: | :---: | :------: |

> |            [ATSS](https://github.com/open-mmlab/mmdetection/tree/master/configs/atss)            |   ObjectDetection    |      Y      |    Y     |  N   |   N   |    Y     |

Reviewer: 'ObjectDetection' or 'Object Detection'? Does `ObjectDetection` has specific meanings?

Author: no specific meaning, just indicates a task for detecting object

```{tip}
When converting mmdet models to tensorrt models, --device should be set to "cuda"
```

## Model specification

Before moving on to model inference chapter, let's know more about the converted model structure which is very important for model inference.

The converted model locates in the working directory like `mmdeploy_models/mmdet/ort` in the previous example. It includes:

```
mmdeploy_models/mmdet/ort
├── deploy.json
├── detail.json
├── end2end.onnx
└── pipeline.json
```

in which,

- **end2end.onnx**: backend model which can be inferred by ONNX Runtime
- **deploy.json**: meta information about backend model
- **pipeline.json**: inference pipeline of mmdeploy SDK
- **detail.json**: conversion parameters

The whole package **mmdeploy_models/mmdet/ort** is defined as **mmdeploy SDK model**, i.e., **mmdeploy SDK model** includes both backend model and inference meta information.

## Model inference

### Backend model inference

> MMDeploy provides a unified API named as `inference_model` to inference model, making all inference backends API transparent to users.

Reviewer 2: The code below is not using inference_model

Author: Done. This line is removed.
